cli/src/cmd/cast/run.rs

```diff
@@ -0,0 +1,210 @@
+    types::H256,
+};
+use forge::{
+    debug::DebugArena,
+    executor::{builder::Backend, opts::EvmOpts, DeployResult, ExecutorBuilder, RawCallResult},
+    trace::{identifier::EtherscanIdentifier, CallTraceArena, CallTraceDecoderBuilder, TraceKind},
+};
+use foundry_config::Config;
+use foundry_utils::RuntimeOrHandle;
+use std::{
+    collections::{BTreeMap, HashMap},
+    str::FromStr,
+    time::Duration,
+};
+use ui::{TUIExitReason, Tui, Ui};
+
+#[derive(Debug, Clone, Parser)]
+pub struct RunArgs {
+    #[clap(help = "The transaction hash.")]
+    tx: String,
+    #[clap(short, long, env = "ETH_RPC_URL")]
+    rpc_url: String,
+    #[clap(long, short = 'd', help = "Debugs the transaction.")]
+    debug: bool,
+    #[clap(
+        long,
+        short = 'q',
+        help = "Executes the transaction only with the state from the previous block. May result in different results than the live execution!"
+    )]
+    quick: bool,
+    #[clap(
+        long,
+        help = "Labels address in the trace. 0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045:vitalik.eth"
+    )]
+    label: Vec<String>,
+}
+
+impl Cmd for RunArgs {
+    type Output = ();
+    fn run(self) -> eyre::Result<Self::Output> {
+        RuntimeOrHandle::new().block_on(self.run_tx())
+    }
+}
+
+impl RunArgs {
+    async fn run_tx(self) -> eyre::Result<()> {
+        let figment = Config::figment();
+        let mut evm_opts = figment.extract::<EvmOpts>()?;
+        let config = Config::from_provider(figment).sanitized();
+
+        let provider =
+            Provider::try_from(self.rpc_url.as_str()).expect("could not instantiate provider");
+
+        if let Some(tx) =
+            provider.get_transaction(H256::from_str(&self.tx).expect("invalid tx hash")).await?
+        {
+            let tx_block_number = tx.block_number.expect("no block number").as_u64();
+            let tx_hash = tx.hash();
+            evm_opts.fork_url = Some(self.rpc_url);
+            evm_opts.fork_block_number = Some(tx_block_number - 1);
+
+            // Set up the execution environment
+            let env = evm_opts.evm_env().await;
+            let db =
+                Backend::new(utils::get_fork(&evm_opts, &config.rpc_storage_caching), &env).await;
+
+            let builder = ExecutorBuilder::new()
+                .with_config(env)
+                .with_spec(crate::utils::evm_spec(&config.evm_version));
+
+            let mut executor = builder.build(db);
+
+            // Set the state to the moment right before the transaction
+            if !self.quick {
+                println!("Executing previous transactions from the block.");
+
+                let block_txes = provider.get_block_with_txs(tx_block_number).await?;
+
+                for past_tx in block_txes.unwrap().transactions.into_iter() {
+                    if past_tx.hash().eq(&tx_hash) {
+                        break
+                    }
+
+                    executor.set_gas_limit(tx.gas);
```

Comment:
Looks like this is setting the gas limit of past txs to our txs gas limit

```suggestion
                    executor.set_gas_limit(past_tx.gas);
```